src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java

```diff
@@ -2218,7 +2218,6 @@ private static void addAstChild(DetailAstPair pair, DetailAstImpl ast) {
                     pair.child.setNextSibling(ast);
                 }
                 pair.child = ast;
-                pair.advanceChildToEnd();
```

Comment:
ANTLR Regression: http://rveach.no-ip.org/checkstyle/regression/337/

`advanceChildToEnd` does a constant `child.getNextSibling()`, which equates to `pair.child.getNextSibling()` which equates to `ast.getNextSibling()`.

The `ast` given to this method never has any siblings, so the method call always does absolutely nothing. I ran all tests and they should the same information, no sibling.

Reply:
> The ast given to this method never has any siblings.

```
➜  src cat Test.java
public class Test {
    public static void main(String[] args) {
        (new B()).<String>m();
    }

    static class B<T> {
        void m() {
        }
    }
}

```
Hacked checkstyle:
```
➜  checkstyle git:(master) ✗ git diff | cat
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java b/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java
index d01a45bb4..56f583797 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java
@@ -2218,6 +2218,10 @@ public final class JavaAstVisitor extends JavaLanguageParserBaseVisitor<DetailAs
                     pair.child.setNextSibling(ast);
                 }
                 pair.child = ast;
+                if (ast.getNextSibling() != null) {
+                    System.out.println("AST: " + ast);
+                    System.out.println("AST sibling:" + ast.getNextSibling());
+                }
                 pair.advanceChildToEnd();
             }
         }


```

result:
```
AST: ([3x8]
AST sibling:new[3x9]
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
`--CLASS_DEF -> CLASS_DEF [1:0]
    |--MODIFIERS -> MODIFIERS [1:0]
    |   `--LITERAL_PUBLIC -> public [1:0]
    |--LITERAL_CLASS -> class [1:7]
    |--IDENT -> Test [1:13]
    `--OBJBLOCK -> OBJBLOCK [1:18]
        |--LCURLY -> { [1:18]
        |--METHOD_DEF -> METHOD_DEF [2:4]
        |   |--MODIFIERS -> MODIFIERS [2:4]
        |   |   |--LITERAL_PUBLIC -> public [2:4]
        |   |   `--LITERAL_STATIC -> static [2:11]
        |   |--TYPE -> TYPE [2:18]
        |   |   `--LITERAL_VOID -> void [2:18]
        |   |--IDENT -> main [2:23]
        |   |--LPAREN -> ( [2:27]
        |   |--PARAMETERS -> PARAMETERS [2:28]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [2:28]
        |   |       |--MODIFIERS -> MODIFIERS [2:28]
        |   |       |--TYPE -> TYPE [2:28]
        |   |       |   |--IDENT -> String [2:28]
        |   |       |   `--ARRAY_DECLARATOR -> [ [2:34]
        |   |       |       `--RBRACK -> ] [2:35]
        |   |       `--IDENT -> args [2:37]
        |   |--RPAREN -> ) [2:41]
        |   `--SLIST -> { [2:43]
        |       |--EXPR -> EXPR [3:27]
        |       |   `--METHOD_CALL -> ( [3:27]
        |       |       |--DOT -> . [3:17]
        |       |       |   |--LPAREN -> ( [3:8]
        |       |       |   |--LITERAL_NEW -> new [3:9]
        |       |       |   |   |--IDENT -> B [3:13]
        |       |       |   |   |--LPAREN -> ( [3:14]
        |       |       |   |   |--ELIST -> ELIST [3:15]
        |       |       |   |   `--RPAREN -> ) [3:15]
        |       |       |   |--RPAREN -> ) [3:16]
        |       |       |   |--TYPE_ARGUMENTS -> TYPE_ARGUMENTS [3:18]
        |       |       |   |   |--GENERIC_START -> < [3:18]
        |       |       |   |   |--TYPE_ARGUMENT -> TYPE_ARGUMENT [3:19]
        |       |       |   |   |   `--IDENT -> String [3:19]
        |       |       |   |   `--GENERIC_END -> > [3:25]
        |       |       |   `--IDENT -> m [3:26]
        |       |       |--ELIST -> ELIST [3:28]
        |       |       `--RPAREN -> ) [3:28]
        |       |--SEMI -> ; [3:29]
        |       `--RCURLY -> } [4:4]
        |--CLASS_DEF -> CLASS_DEF [6:4]
        |   |--MODIFIERS -> MODIFIERS [6:4]
        |   |   `--LITERAL_STATIC -> static [6:4]
        |   |--LITERAL_CLASS -> class [6:11]
        |   |--IDENT -> B [6:17]
        |   |--TYPE_PARAMETERS -> TYPE_PARAMETERS [6:18]
        |   |   |--GENERIC_START -> < [6:18]
        |   |   |--TYPE_PARAMETER -> TYPE_PARAMETER [6:19]
        |   |   |   `--IDENT -> T [6:19]
        |   |   `--GENERIC_END -> > [6:20]
        |   `--OBJBLOCK -> OBJBLOCK [6:22]
        |       |--LCURLY -> { [6:22]
        |       |--METHOD_DEF -> METHOD_DEF [7:8]
        |       |   |--MODIFIERS -> MODIFIERS [7:8]
        |       |   |--TYPE -> TYPE [7:8]
        |       |   |   `--LITERAL_VOID -> void [7:8]
        |       |   |--IDENT -> m [7:13]
        |       |   |--LPAREN -> ( [7:14]
        |       |   |--PARAMETERS -> PARAMETERS [7:15]
        |       |   |--RPAREN -> ) [7:15]
        |       |   `--SLIST -> { [7:17]
        |       |       `--RCURLY -> } [8:8]
        |       `--RCURLY -> } [9:4]
        `--RCURLY -> } [10:0]


```

Reply:
I think it would be better to break this off into another PR.

Reply:
I would rather not recreate the suppression file I removed.

From your example, the reason why it doesn't matter is because we call `DetailAstPair.makeAstRoot` right after the affected `DetailAstPair.addAstChild`. `makeAstRoot` doesn't do anything with `pair.child` , only `pair.root`.

It seems to me, we need 2 `DetailAstPair.addAstChild` in a row, with the first one having an extra sibling.

Reply:
> It seems to me, we need 2 DetailAstPair.addAstChild in a row, with the first one having an extra sibling in the AST passed to it.

No, this will notwork either. I honestly do not like the `DetailAstPair` class at all, this is our only remnant of ANTLR2. If we can begin to clean up code here, let's do it.